hyper-capi/include/hyper.h

```diff
@@ -0,0 +1,318 @@
+#ifndef _HYPER_H
+#define _HYPER_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#include <stdint.h>
+#include <stddef.h>
+
+typedef struct hyper_clientconn hyper_clientconn;
+
+typedef struct hyper_clientconn_options hyper_clientconn_options;
+
+typedef struct hyper_io hyper_io;
+
+typedef struct hyper_request hyper_request;
+
+typedef struct hyper_response hyper_response;
+
+typedef struct hyper_headers hyper_headers;
+
+typedef struct hyper_body hyper_body;
+
+typedef struct hyper_buf hyper_buf;
+
+typedef struct hyper_task hyper_task;
+
+typedef struct hyper_context hyper_context;
+
+typedef struct hyper_waker hyper_waker;
+
+typedef struct hyper_executor hyper_executor;
+
+typedef enum {
+    HYPER_OK = 0,
+    HYPERE_KABOOM
+} hyper_code;
+
+typedef enum {
+	HYPER_ITER_CONTINUE,
+	HYPER_ITER_BREAK
+} hyper_iter_step;
+
+typedef enum {
+	HYPER_POLL_READY,
+	HYPER_POLL_PENDING
+} hyper_poll;
+
+// HTTP ClientConn
+
+// Starts an HTTP client connection handshake using the provided IO transport
+// and options.
+//
+// Both the `io` and the `options` are consumed in this function call.
+//
+// The returned `hyper_task *` must be polled with an executor until the
+// handshake completes, at which point the value can be taken.
+hyper_task *hyper_clientconn_handshake(hyper_io *io, hyper_clientconn_options *options);
+
+// Send a request on the client connection.
+//
+// Returns a task that needs to be polled until it is ready. When ready, the
+// task yields a `hyper_response *`.
+hyper_task *hyper_clientconn_send(hyper_clientconn *client, hyper_request *request);
+
+// Free a `hyper_clientconn *`.
+void hyper_clientconn_free(hyper_clientconn *client);
+
+// Creates a new set of HTTP clientconn options to be used in a handshake.
+hyper_clientconn_options *hyper_clientconn_options_new(void);
+
+// Set the client background task executor.
```

Comment:
Hmm. What is this used for? We don't have any hooks for setting timeouts (for example), so I'm not sure what need there would be for a background executor at this time.

Reply:
hyper puts the "connection" itself into a separate task that is always polled, to handle hang up detection, and any other meta stuff such as when using HTTP/2. The `clientconn` that is returned is a handle to send requests into that task.